File: hdf-engine/go/merge_test.go
package hdfengine

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
	"time"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
	validators "github.com/mitre/hdf-libs/hdf-validators/go/v3"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// loadMergeFixture reads one of the shared merge fixtures — committed copies of
// real converter expected output (gosec real, ZAP webgoat, grype tensorflow).
// test/merge.test.ts reads the SAME files and asserts the SAME expectations, so
// the two Merge implementations are held to one cross-language contract.
func loadMergeFixture(t *testing.T, name string) hdf.HDFResults {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("..", "testdata", name))
	require.NoError(t, err)
	var results hdf.HDFResults
	require.NoError(t, json.Unmarshal(data, &results))
	return results
}

func threeScanners(t *testing.T) []MergeSource {
	t.Helper()
	return []MergeSource{
		{Name: "gosec.hdf.json", Doc: loadMergeFixture(t, "merge-gosec.json")},
		{Name: "zap.hdf.json", Doc: loadMergeFixture(t, "merge-zap.json")},
		{Name: "grype.hdf.json", Doc: loadMergeFixture(t, "merge-grype.json")},
	}
}

func baselineNames(r hdf.HDFResults) []string {
	out := make([]string, 0, len(r.Baselines))
	for _, b := range r.Baselines {
		out = append(out, b.Name)
	}
	return out
}

// TestMerge_OneBaselinePerInputPrefixedByTool is js1nv.3's first failing test:
// merging gosec + ZAP + grype yields one baseline per input baseline (1 + 4 + 1),
// each renamed `<tool>/<original>` per ADR-0016 §2, with every requirement
// carried and the output valid against hdf-results.
func TestMerge_OneBaselinePerInputPrefixedByTool(t *testing.T) {
	merged, warnings, err := Merge(threeScanners(t))
	require.NoError(t, err)
	assert.Empty(t, warnings, "distinct tools and names produce no warnings")

	assert.Equal(t, []string{
		"gosec/gosec Scan",
		"owasp zap/OWASP ZAP Scan: ciscobinary.openh264.org",
		"owasp zap/OWASP ZAP Scan: code.jquery.com",
		"owasp zap/OWASP ZAP Scan: detectportal.firefox.com",
		"owasp zap/OWASP ZAP Scan: mymac.com",
		"grype/tensorflow/tensorflow:latest",
	}, baselineNames(merged))

	total := 0
	for _, b := range merged.Baselines {
		total += len(b.Requirements)
	}
	assert.Equal(t, 57, total, "3 + 28 + 26 requirements, none dropped or deduplicated")

	data, err := json.Marshal(merged)
	require.NoError(t, err)
	res := validators.ValidateResults(data)
	assert.True(t, res.Valid, "merged document must validate as hdf-results: %+v", res.Errors)
}

// TestMerge_ProvenanceLabels (ADR-0016 §3): every merged baseline carries
// tool / toolVersion / sourceDocument; labels the source already had survive.
func TestMerge_ProvenanceLabels(t *testing.T) {
	merged, _, err := Merge(threeScanners(t))
	require.NoError(t, err)

	gosec := merged.Baselines[0]
	assert.Equal(t, map[string]string{"tool": "gosec", "toolVersion": "dev", "sourceDocument": "gosec.hdf.json"}, gosec.Labels)
	// The gosec baseline's own extensions ride along untouched.
	assert.Contains(t, gosec.Extensions, "gosec")

	zap := merged.Baselines[1]
	assert.Equal(t, map[string]string{
		"component": "ciscobinary.openh264.org", // pre-existing label, preserved
		"tool":      "owasp zap", "toolVersion": "2.7.0", "sourceDocument": "zap.hdf.json",
	}, zap.Labels)

	grype := merged.Baselines[5]
	assert.Equal(t, map[string]string{"tool": "grype", "toolVersion": "0.79.3", "sourceDocument": "grype.hdf.json"}, grype.Labels)
}

// TestMerge_Root (ADR-0016 §4): generator names the merger at the engine
// version; tool is omitted; timestamp is the latest input; components are the
// union (none of these fixtures carries a componentId, so all five are kept);
// each input's root provenance is preserved verbatim under extensions.
func TestMerge_Root(t *testing.T) {
	merged, _, err := Merge(threeScanners(t))
	require.NoError(t, err)

	require.NotNil(t, merged.Generator)
	assert.Equal(t, hdf.Generator{Name: "hdf-merge", Version: Version()}, *merged.Generator)
	assert.Nil(t, merged.Tool)
	require.NotNil(t, merged.Timestamp)
	assert.Equal(t, "2026-07-12T22:56:36.173673Z", merged.Timestamp.UTC().Format("2006-01-02T15:04:05.999999999Z07:00"), "gosec's timestamp is the latest of the three")
	assert.Len(t, merged.Components, 5, "4 ZAP sites + 1 grype image, no componentIds → all kept")

	ext, ok := merged.Extensions["hdf-merge"].(map[string]any)
	require.True(t, ok, "extensions[hdf-merge] must be present: %v", merged.Extensions)
	assert.Equal(t, Version(), ext["version"])
	sources, ok := ext["sources"].([]any)
	require.True(t, ok)
	require.Len(t, sources, 3)
	first, ok := sources[0].(map[string]any)
	require.True(t, ok)
	assert.EqualValues(t, 0, first["index"])
	assert.Equal(t, "gosec.hdf.json", first["name"])
	assert.Equal(t, map[string]any{"name": "gosec", "version": "dev"}, first["tool"])
	assert.Equal(t, map[string]any{"name": "gosec-to-hdf", "version": "1.0.0"}, first["generator"])
	assert.Equal(t, "2026-07-12T22:56:36.173673Z", first["timestamp"])
	_, hasRunner := first["runner"]
	assert.False(t, hasRunner, "an absent input field is absent, never synthesized")
}

// TestMerge_Deterministic (ADR-0016 §4): the same inputs in the same order
// serialize to byte-identical JSON on every run.
func TestMerge_Deterministic(t *testing.T) {
	a, _, err := Merge(threeScanners(t))
	require.NoError(t, err)
	b, _, err := Merge(threeScanners(t))
	require.NoError(t, err)
	ja, err := json.Marshal(a)
	require.NoError(t, err)
	jb, err := json.Marshal(b)
	require.NoError(t, err)
	assert.Equal(t, string(ja), string(jb))
}

// TestMerge_WarnsOnCollisionsAndOverwrites (ADR-0016 §2, §3): two inputs from
// the same tool with the same baseline name still collide after prefixing —
// both are kept and a warning names both positions. Merging an already-merged
// document overwrites its provenance labels, with a warning per label, and its
// prefix falls back to generator.name because a merged root carries no tool.
func TestMerge_WarnsOnCollisionsAndOverwrites(t *testing.T) {
	g := loadMergeFixture(t, "merge-gosec.json")
	merged, warnings, err := Merge([]MergeSource{{Name: "a.json", Doc: g}, {Name: "b.json", Doc: g}})
	require.NoError(t, err)
	assert.Equal(t, []string{"gosec/gosec Scan", "gosec/gosec Scan"}, baselineNames(merged), "collisions are kept, never dropped or renamed further")
	require.Len(t, warnings, 1)
	assert.Equal(t, MergeWarning{Kind: WarnDuplicateBaselineName, Name: "gosec/gosec Scan", Indices: []int{0, 1}}, warnings[0])

	again, warnings, err := Merge([]MergeSource{{Name: "merged.json", Doc: merged}})
	require.NoError(t, err)
	assert.Equal(t, []string{"hdf-merge/gosec/gosec Scan", "hdf-merge/gosec/gosec Scan"}, baselineNames(again))
	kinds := map[MergeWarningKind]int{}
	for _, w := range warnings {
		kinds[w.Kind]++
	}
	assert.Equal(t, 1, kinds[WarnDuplicateBaselineName], "the collision is reported again, exactly once")
	assert.Equal(t, 6, kinds[WarnLabelOverwritten], "tool, toolVersion and sourceDocument overwritten on each of the two baselines")
	assert.Equal(t, "hdf-merge", again.Baselines[0].Labels["tool"])
	assert.Equal(t, "merged.json", again.Baselines[0].Labels["sourceDocument"])
}

// TestMerge_Errors: no sources is an error, not an empty document.
func TestMerge_Errors(t *testing.T) {
	_, _, err := Merge(nil)
	require.Error(t, err)
	_, _, err = Merge([]MergeSource{})
	require.Error(t, err)
}

// TestMerge_WritesForLiveVerification is the Gate-18 hook: when
// HDF_MERGE_LIVE_OUT names a path, it writes the three-scanner merge there so
// the shipped `hdf` binary can be run against real merged output (validate,
// query, list). Skipped otherwise — it asserts nothing the tests above do not.
func TestMerge_WritesForLiveVerification(t *testing.T) {
	out := os.Getenv("HDF_MERGE_LIVE_OUT")
	if out == "" {
		t.Skip("set HDF_MERGE_LIVE_OUT=<path> to write the merged document for live verification")
	}
	merged, _, err := Merge(threeScanners(t))
	require.NoError(t, err)
	data, err := json.MarshalIndent(merged, "", "  ")
	require.NoError(t, err)
	require.NoError(t, os.WriteFile(out, data, 0o600))
}

// The tests below pin the ADR-0016 rules the three-scanner happy path cannot
// distinguish (its latest timestamp is also its first input, its fixtures carry
// no componentIds, and every input names a tool). Where a rule needs a document
// shape no committed real fixture has, the input is DERIVED from a real fixture
// in-test — a field cleared or an id assigned — never a fabricated document.

// TestMerge_LatestTimestampIsNotFirst: with grype (2024) first and gosec (2026)
// second, the root timestamp is still gosec's, so "latest" is not "first".
func TestMerge_LatestTimestampIsNotFirst(t *testing.T) {
	merged, _, err := Merge([]MergeSource{
		{Name: "grype.hdf.json", Doc: loadMergeFixture(t, "merge-grype.json")},
		{Name: "gosec.hdf.json", Doc: loadMergeFixture(t, "merge-gosec.json")},
		{Name: "zap.hdf.json", Doc: loadMergeFixture(t, "merge-zap.json")},
	})
	require.NoError(t, err)
	require.NotNil(t, merged.Timestamp)
	assert.Equal(t, "2026-07-12T22:56:36.173673Z", merged.Timestamp.UTC().Format(time.RFC3339Nano))
	assert.Equal(t, "grype/tensorflow/tensorflow:latest", merged.Baselines[0].Name, "input order is preserved")
	assert.Equal(t, "gosec/gosec Scan", merged.Baselines[1].Name)
}

// TestMerge_NoTimestampOnAnyInput: the root timestamp is omitted, not invented.
func TestMerge_NoTimestampOnAnyInput(t *testing.T) {
	g := loadMergeFixture(t, "merge-gosec.json")
	g.Timestamp = nil
	merged, _, err := Merge([]MergeSource{{Name: "g.json", Doc: g}})
	require.NoError(t, err)
	assert.Nil(t, merged.Timestamp)
	src := merged.Extensions["hdf-merge"].(map[string]any)["sources"].([]any)[0].(map[string]any)
	_, has := src["timestamp"]
	assert.False(t, has, "provenance carries no timestamp when the input had none")
}

// TestMerge_PrefixFallbacks: tool.name is trimmed and lower-cased; with no tool
// the generator name is used; with neither the prefix is doc<N>, 0-based, and no
// toolVersion label is written.
func TestMerge_PrefixFallbacks(t *testing.T) {
	spaced := loadMergeFixture(t, "merge-gosec.json")
	name := "  GoSec "
	spaced.Tool.Name = &name

	noTool := loadMergeFixture(t, "merge-gosec.json")
	noTool.Tool = nil // generator gosec-to-hdf remains

	bare := loadMergeFixture(t, "merge-gosec.json")
	bare.Tool = nil
	bare.Generator = nil

	merged, warnings, err := Merge([]MergeSource{
		{Name: "spaced.json", Doc: spaced},
		{Name: "notool.json", Doc: noTool},
		{Name: "bare.json", Doc: bare},
	})
	require.NoError(t, err)
	assert.Equal(t, []string{"gosec/gosec Scan", "gosec-to-hdf/gosec Scan", "doc2/gosec Scan"}, baselineNames(merged))
	assert.Equal(t, "gosec", merged.Baselines[0].Labels["tool"])
	assert.Equal(t, "gosec-to-hdf", merged.Baselines[1].Labels["tool"])
	assert.Equal(t, map[string]string{"tool": "doc2", "sourceDocument": "bare.json"}, merged.Baselines[2].Labels, "no tool → no toolVersion label")
	assert.Empty(t, warnings)
}

// TestMerge_ComponentUnionByID: components sharing a componentId are kept once;
// components without one are all kept. Derived from real documents: the grype
// image component is given a fixed UUID and merged with itself.
func TestMerge_ComponentUnionByID(t *testing.T) {
	id := "11111111-1111-4111-8111-111111111111"
	a := loadMergeFixture(t, "merge-grype.json")
	a.Components[0].ComponentID = &id
	b := loadMergeFixture(t, "merge-grype.json")
	b.Components[0].ComponentID = &id
	zap := loadMergeFixture(t, "merge-zap.json") // four components, no ids

	merged, _, err := Merge([]MergeSource{{Name: "a.json", Doc: a}, {Name: "b.json", Doc: b}, {Name: "zap.json", Doc: zap}})
	require.NoError(t, err)
	require.Len(t, merged.Components, 5, "1 (shared id, kept once) + 4 (no id, all kept)")
	assert.Equal(t, id, *merged.Components[0].ComponentID)
	assert.Nil(t, merged.Components[1].ComponentID)
	data, err := json.Marshal(merged)
	require.NoError(t, err)
	assert.True(t, validators.ValidateResults(data).Valid)
}

// TestMerge_RemergeReplacesProvenance: merging an already-merged document
// replaces tool and sourceDocument, REMOVES the stale toolVersion (the merged
// root has no tool, so a kept version would describe the wrong tool), and
// reports each of the three per baseline with the label named.
func TestMerge_RemergeReplacesProvenance(t *testing.T) {
	first, _, err := Merge([]MergeSource{{Name: "zap.hdf.json", Doc: loadMergeFixture(t, "merge-zap.json")}})
	require.NoError(t, err)
	assert.Equal(t, "2.7.0", first.Baselines[0].Labels["toolVersion"])

	again, warnings, err := Merge([]MergeSource{{Name: "merged.json", Doc: first}})
	require.NoError(t, err)
	for i, b := range again.Baselines {
		_, has := b.Labels["toolVersion"]
		assert.False(t, has, "baseline %d must not keep a stale toolVersion", i)
		assert.Equal(t, "hdf-merge", b.Labels["tool"])
		assert.Equal(t, "merged.json", b.Labels["sourceDocument"])
		assert.Contains(t, b.Labels, "component", "the ZAP site label is preserved through a re-merge")
	}
	got := map[string][]int{}
	for _, w := range warnings {
		require.Equal(t, WarnLabelOverwritten, w.Kind)
		got[w.Label] = append(got[w.Label], w.Indices[0])
	}
	assert.Equal(t, map[string][]int{"tool": {0, 1, 2, 3}, "toolVersion": {0, 1, 2, 3}, "sourceDocument": {0, 1, 2, 3}}, got)
}

// TestMerge_SingleDocumentIsIdentity: apart from the renamed baselines, their
// labels, and the merged root, one input comes out unchanged — requirements,
// baseline metadata, extensions and components are the same values.
func TestMerge_SingleDocumentIsIdentity(t *testing.T) {
	in := loadMergeFixture(t, "merge-zap.json")
	merged, warnings, err := Merge([]MergeSource{{Name: "zap.hdf.json", Doc: in}})
	require.NoError(t, err)
	assert.Empty(t, warnings)
	require.Len(t, merged.Baselines, len(in.Baselines))
	for i := range in.Baselines {
		want := in.Baselines[i]
		got := merged.Baselines[i]
		want.Name, got.Name = "", ""
		want.Labels, got.Labels = nil, nil
		assert.Equal(t, want, got, "baseline %d differs beyond name/labels", i)
	}
	assert.Equal(t, in.Components, merged.Components)
	// The input itself was not mutated.
	assert.Equal(t, "OWASP ZAP Scan: ciscobinary.openh264.org", in.Baselines[0].Name)
	assert.Equal(t, map[string]string{"component": "ciscobinary.openh264.org"}, in.Baselines[0].Labels)
}

File: hdf-engine/test/merge.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import type { HDFResults } from '@mitre/hdf-schema';
import { results as sharedResults } from '@mitre/hdf-fixtures';
import { merge, type MergeSource } from '../src/merge.js';
import { engineVersion } from '../src/index.js';
import { validateResults } from '@mitre/hdf-validators';

// merge-grype/merge-zap live in the shared corpus (@mitre/hdf-fixtures), read
// here and by go/merge_test.go; merge-gosec is still local to this package.
// Both languages read the SAME bytes and assert the SAME expectations.
const testdata = join(dirname(fileURLToPath(import.meta.url)), '..', 'testdata');
const load = (name: string): HDFResults => {
  const raw =
    name === 'merge-grype.json'
      ? sharedResults.mergeGrype.read()
      : name === 'merge-zap.json'
        ? sharedResults.mergeZap.read()
        : readFileSync(join(testdata, name), 'utf-8');
  return JSON.parse(raw) as HDFResults;
};

function threeScanners(): MergeSource[] {
  return [
    { name: 'gosec.hdf.json', doc: load('merge-gosec.json') },
    { name: 'zap.hdf.json', doc: load('merge-zap.json') },
    { name: 'grype.hdf.json', doc: load('merge-grype.json') },
  ];
}

const names = (r: HDFResults): string[] => r.baselines.map((b) => b.name);

// Everything about a baseline except the two fields merge rewrites.
function withoutNameAndLabels(b: HDFResults['baselines'][number]): Omit<typeof b, 'name' | 'labels'> {
  const copy: Partial<typeof b> = { ...b };
  delete copy.name;
  delete copy.labels;
  return copy as Omit<typeof b, 'name' | 'labels'>;
}

describe('hdf-engine merge — cross-language parity with go/merge_test.go', () => {
  it('one baseline per input baseline, renamed <tool>/<original>, every requirement kept', () => {
    const { results, warnings } = merge(threeScanners());
    expect(warnings).toEqual([]);
    expect(names(results)).toEqual([
      'gosec/gosec Scan',
      'owasp zap/OWASP ZAP Scan: ciscobinary.openh264.org',
      'owasp zap/OWASP ZAP Scan: code.jquery.com',
      'owasp zap/OWASP ZAP Scan: detectportal.firefox.com',
      'owasp zap/OWASP ZAP Scan: mymac.com',
      'grype/tensorflow/tensorflow:latest',
    ]);
    expect(results.baselines.reduce((n, b) => n + b.requirements.length, 0)).toBe(57);
  });

  it('provenance labels: tool / toolVersion / sourceDocument, existing labels preserved', () => {
    const { results } = merge(threeScanners());
    expect(results.baselines[0].labels).toEqual({ tool: 'gosec', toolVersion: 'dev', sourceDocument: 'gosec.hdf.json' });
    expect(results.baselines[0].extensions).toHaveProperty('gosec');
    expect(results.baselines[1].labels).toEqual({
      component: 'ciscobinary.openh264.org',
      tool: 'owasp zap',
      toolVersion: '2.7.0',
      sourceDocument: 'zap.hdf.json',
    });
    expect(results.baselines[5].labels).toEqual({ tool: 'grype', toolVersion: '0.79.3', sourceDocument: 'grype.hdf.json' });
  });

  it('root: generator hdf-merge@engineVersion, no tool, latest timestamp, component union, sources verbatim', () => {
    const { results } = merge(threeScanners());
    expect(results.generator).toEqual({ name: 'hdf-merge', version: engineVersion });
    expect(results.tool).toBeUndefined();
    expect(results.timestamp).toBe('2026-07-12T22:56:36.173673Z');
    expect(results.components).toHaveLength(5);
    const ext = results.extensions?.['hdf-merge'] as { version: string; sources: Record<string, unknown>[] };
    expect(ext.version).toBe(engineVersion);
    expect(ext.sources).toHaveLength(3);
    expect(ext.sources[0]).toEqual({
      index: 0,
      name: 'gosec.hdf.json',
      tool: { name: 'gosec', version: 'dev' },
      generator: { name: 'gosec-to-hdf', version: '1.0.0' },
      timestamp: '2026-07-12T22:56:36.173673Z',
    });
  });

  it('is deterministic: identical inputs serialize identically', () => {
    const a = JSON.stringify(merge(threeScanners()).results);
    const b = JSON.stringify(merge(threeScanners()).results);
    expect(a).toBe(b);
  });

  it('warns on name collisions and label overwrites; merged roots prefix by generator', () => {
    const g = load('merge-gosec.json');
    const first = merge([
      { name: 'a.json', doc: g },
      { name: 'b.json', doc: g },
    ]);
    expect(names(first.results)).toEqual(['gosec/gosec Scan', 'gosec/gosec Scan']);
    expect(first.warnings).toEqual([{ kind: 'duplicate-baseline-name', name: 'gosec/gosec Scan', indices: [0, 1] }]);

    const again = merge([{ name: 'merged.json', doc: first.results }]);
    expect(names(again.results)).toEqual(['hdf-merge/gosec/gosec Scan', 'hdf-merge/gosec/gosec Scan']);
    expect(again.warnings.filter((w) => w.kind === 'duplicate-baseline-name')).toHaveLength(1);
    expect(again.warnings.filter((w) => w.kind === 'label-overwritten')).toHaveLength(6);
    expect(again.results.baselines[0].labels?.tool).toBe('hdf-merge');
    expect(again.results.baselines[0].labels?.sourceDocument).toBe('merged.json');
  });

  it('rejects an empty source list', () => {
    expect(() => merge([])).toThrow();
  });
});

// Rules the three-scanner happy path cannot distinguish (parity with the Go
// tests of the same names). Inputs that need a shape no committed real fixture
// has are DERIVED from a real fixture in-test — never fabricated.
describe('hdf-engine merge — rule coverage, parity with go/merge_test.go', () => {
  it('latest timestamp is not the first input', () => {
    const { results } = merge([
      { name: 'grype.hdf.json', doc: load('merge-grype.json') },
      { name: 'gosec.hdf.json', doc: load('merge-gosec.json') },
      { name: 'zap.hdf.json', doc: load('merge-zap.json') },
    ]);
    expect(results.timestamp).toBe('2026-07-12T22:56:36.173673Z');
    expect(results.baselines[0].name).toBe('grype/tensorflow/tensorflow:latest');
    expect(results.baselines[1].name).toBe('gosec/gosec Scan');
  });

  it('no timestamp on any input → omitted, not invented', () => {
    const g = load('merge-gosec.json');
    delete g.timestamp;
    const { results } = merge([{ name: 'g.json', doc: g }]);
    expect(results.timestamp).toBeUndefined();
    const ext = results.extensions?.['hdf-merge'] as { sources: Record<string, unknown>[] };
    expect(ext.sources[0]).not.toHaveProperty('timestamp');
  });

  it('prefix fallbacks: trimmed lower-case tool → generator → doc<N>', () => {
    const spaced = load('merge-gosec.json');
    spaced.tool = { ...spaced.tool, name: '  GoSec ' };
    const noTool = load('merge-gosec.json');
    delete noTool.tool;
    const bare = load('merge-gosec.json');
    delete bare.tool;
    delete bare.generator;
    const { results, warnings } = merge([
      { name: 'spaced.json', doc: spaced },
      { name: 'notool.json', doc: noTool },
      { name: 'bare.json', doc: bare },
    ]);
    expect(names(results)).toEqual(['gosec/gosec Scan', 'gosec-to-hdf/gosec Scan', 'doc2/gosec Scan']);
    expect(results.baselines[0].labels?.tool).toBe('gosec');
    expect(results.baselines[1].labels?.tool).toBe('gosec-to-hdf');
    expect(results.baselines[2].labels).toEqual({ tool: 'doc2', sourceDocument: 'bare.json' });
    expect(warnings).toEqual([]);
  });

  it('components sharing a componentId are kept once; those without one are all kept', () => {
    const id = '11111111-1111-4111-8111-111111111111';
    const a = load('merge-grype.json');
    a.components![0].componentId = id;
    const b = load('merge-grype.json');
    b.components![0].componentId = id;
    const { results } = merge([
      { name: 'a.json', doc: a },
      { name: 'b.json', doc: b },
      { name: 'zap.json', doc: load('merge-zap.json') },
    ]);
    expect(results.components).toHaveLength(5);
    expect(results.components![0].componentId).toBe(id);
    expect(results.components![1].componentId).toBeUndefined();
    expect(validateResults(results).valid).toBe(true);
  });

  it('re-merge replaces tool/sourceDocument, removes the stale toolVersion, and names each label', () => {
    const first = merge([{ name: 'zap.hdf.json', doc: load('merge-zap.json') }]).results;
    expect(first.baselines[0].labels?.toolVersion).toBe('2.7.0');
    const { results, warnings } = merge([{ name: 'merged.json', doc: first }]);
    for (const b of results.baselines) {
      expect(b.labels).not.toHaveProperty('toolVersion');
      expect(b.labels?.tool).toBe('hdf-merge');
      expect(b.labels?.sourceDocument).toBe('merged.json');
      expect(b.labels).toHaveProperty('component');
    }
    const byLabel: Record<string, number[]> = {};
    for (const w of warnings) {
      expect(w.kind).toBe('label-overwritten');
      (byLabel[w.label!] ??= []).push(w.indices[0]);
    }
    expect(byLabel).toEqual({ tool: [0, 1, 2, 3], toolVersion: [0, 1, 2, 3], sourceDocument: [0, 1, 2, 3] });
  });

  it('a single document is the identity apart from names, labels and the root; the input is not mutated', () => {
    const input = load('merge-zap.json');
    const { results, warnings } = merge([{ name: 'zap.hdf.json', doc: input }]);
    expect(warnings).toEqual([]);
    expect(results.baselines).toHaveLength(input.baselines.length);
    input.baselines.forEach((want, i) => {
      expect(withoutNameAndLabels(results.baselines[i])).toEqual(withoutNameAndLabels(want));
    });
    expect(results.components).toEqual(input.components);
    expect(input.baselines[0].name).toBe('OWASP ZAP Scan: ciscobinary.openh264.org');
    expect(input.baselines[0].labels).toEqual({ component: 'ciscobinary.openh264.org' });
  });

  it('the three-scanner merge validates against hdf-results (TS validators)', () => {
    const { results } = merge(threeScanners());
    const res = validateResults(results);
    expect(res.valid).toBe(true);
  });
});
